inst/include/dust/filter_state.hpp

```diff
@@ -80,74 +66,245 @@ class filter_trajectories {
+    // TODO openmp here?
+    for (size_t i = 0; i < this->n_data_ + 1; ++i) {
+      for (size_t j = 0; j < this->n_particles_; ++j) {
+        for (size_t k = 0; k < this->n_state_; ++k) {
+          destride_history[i * (this->n_particles_ * this->n_state_) +
+                           j * this->n_state_ +
+                           k] = history_host[i * (this->n_particles_ * this->n_state_) +
+                                             j +
+                                             k * (this->n_particles_)];
+        }
+      }
+    }
+    return destride_history;
+  }
+};
+
+template <typename real_t>
+class filter_snapshots {
+public:
+  filter_snapshots() {
+  }
+
+  bool is_snapshot_step(size_t step) {
+    return offset_ < n_steps_ && steps_[offset_] == step;
   }
 
   void advance() {
     offset_++;
   }
 
+protected:
+  size_t n_state_;
+  size_t n_particles_;
+  size_t n_steps_;
+  size_t offset_;
+  std::vector<size_t> steps_;
+};
+
+template <typename real_t>
+class filter_snapshots_host : public filter_snapshots<real_t> {
+public:
+  filter_snapshots_host() {
+  }
+
+  void resize(size_t n_state, size_t n_particles, std::vector<size_t> steps) {
+    this->n_state_ = n_state;
+    this->n_particles_ = n_particles;
+    this->n_steps_ = steps.size();
+    this->offset_ = 0;
+    this->steps_ = steps;
+    state_.resize(this->n_state_ * this->n_particles_ * this->n_steps_);
+  }
+
   size_t size() const {
     return state_.size();
   }
 
+  typename std::vector<real_t>::iterator value_iterator() {
+    return state_.begin() + this->offset_ * this->n_state_ * this->n_particles_;
+  }
+
   template <typename Iterator>
   void history(Iterator dest) const {
     std::copy(state_.begin(), state_.end(), dest);
   }
 
 private:
-  size_t n_state_;
-  size_t n_particles_;
-  size_t n_steps_;
-  size_t offset_;
-  std::vector<size_t> steps_;
   std::vector<real_t> state_;
 };
 
 template <typename real_t>
-struct filter_state {
-  filter_trajectories<real_t> trajectories;
-  filter_snapshots<real_t> snapshots;
+class filter_snapshots_device : public filter_snapshots<real_t> {
+public:
+  filter_snapshots_device() {
+  }
+
+  void resize(size_t n_state, size_t n_particles, std::vector<size_t> steps) {
+    this->n_state_ = n_state;
+    this->n_particles_ = n_particles;
+    this->n_steps_ = steps.size();
+    this->offset_ = 0;
+    this->steps_ = steps;
+    state_ = dust::device_array<real_t>(this->n_state_ * this->n_particles_ * this->n_steps_);
+  }
+
+  size_t size() const {
+    return state_.size();
+  }
+
+  dust::device_array<real_t> &state() {
+    return state_;
+  }
+
+  size_t value_offset() {
+    return this->offset_ * this->n_state_ * this->n_particles_;
+  }
+
+  template <typename Iterator>
+  void history(Iterator dest) const {
+    // Copy from D->H
+    std::vector<real_t> state_host(size());
+    state_.get_array(state_host);
+    // Destride and copy into iterator
+    // TODO: openmp here? collapse(2)
```

Comment:
does this show up in the profile at all?

Reply:
Not sure, we'll have to check with sircovid. Options to deal with this if it is slow would be to either:
- Add an openmp block and see if it helps
- Additionally use std::async or similar to compute this as the next step of the filter is running
- Write a CUDA version of this (similar to the scatter kernel). This would be nicest combined with the particle_ancestry function, which we may want to redo so it's not run backwards after we have the final step (if that's possible)